Expose missing walk/bike steps fields in GraphQL

binh-dam-ibigroup opened the pull request:
### Summary

This PR adds GraphQL fields for walking/biking steps that are counterparts to fields found when using the `plan` endpoint on the REST API but not currently exposed in GraphQL. The fields include relative/absolute (cardinal) direction (which are implemented as enumerated types), street name, etc. needed to display complete turn-by-turn instructions for walking/biking legs.

### Unit tests

- A unit test is added to cover a new mapping function converting a `StreetNote` object to a `TransitAlert` object.
- Manual testing is performed using a GraphQL client, querying an itinerary using Portland, OR OSM and graph.

### Documentation

- Descriptions of the new GraphQL types and fields are added to the GraphQL schema.

leonardehrenfried commented on src/ext/java/org/opentripplanner/ext/legacygraphqlapi/generated/LegacyGraphQLDataFetchers.java:
```diff
@@ -1093,13 +1093,31 @@ public interface LegacyGraphQLServiceTimeRange {
 
     public DataFetcher<Double> lon();
+
+    public DataFetcher<Object> relativeDirection();
```
I'm wondering why the code generator doesn't use the generated enum type here.

@hannesj Do you know?

hannesj replied:
You need to add it to the `src/ext/java/org/opentripplanner/ext/legacygraphqlapi/generated/graphql-codegen.yml`

binh-dam-ibigroup replied:
Enums elsewhere (e.g. `RouteType`) get mapped to `String`, is that the recommended way to do it?

leonardehrenfried replied:
I think the best way is to map it to the generated type, i.e. `LegacyGraphQLRelativeDirection`. 

Take the example of the form factor.

It's configured like this: https://github.com/opentripplanner/OpenTripPlanner/blob/dbf6c5eeabf8d80bcd99652f578afdbe28137a8e/src/ext/java/org/opentripplanner/ext/legacygraphqlapi/generated/graphql-codegen.yml#L92

And then the internal type is mapped to the API type like this: https://github.com/opentripplanner/OpenTripPlanner/blob/9c9dd613489a348d2381acdcbeab8f86589154d7/src/ext/java/org/opentripplanner/ext/legacygraphqlapi/datafetchers/LegacyGraphQLRentalVehicleTypeImpl.java#L13

If we do that we get a compiler error if we changed the internal type but have not type in schema.graphql.

WDYT?

hannesj replied:
> WDYT?

👍

leonardehrenfried commented:
@binh-dam-ibigroup I've taken the liberty to change the implementation so it addresses the feedback given in this PR. Hope this is ok.

I've you have questions, let us know.

binh-dam-ibigroup commented:
> @binh-dam-ibigroup I've taken the liberty to change the implementation so it addresses the feedback given in this PR. Hope this is ok.
> 
> I've you have questions, let us know.

Sure. Thanks for the feedback.

hannesj commented on src/ext-test/java/org/opentripplanner/ext/legacygraphqlapi/LegacyGraphQLstepImplTest.java:
```diff
@@ -0,0 +1,39 @@
+    // Truncate instant to seconds because {@link TimePeriod::new} takes seconds, not milliseconds.
+    Instant startInstant = startDate.toInstant().truncatedTo(ChronoUnit.SECONDS);
+    Instant endInstant = startInstant.plusSeconds(3600);
+    note.effectiveEndDate = Date.from(endInstant);
```
We could refactor these to use `Instant` or `ZonedDateTime`, but can do that in a separate PR

hannesj commented on src/ext/java/org/opentripplanner/ext/legacygraphqlapi/datafetchers/LegacyGraphQLstepImpl.java:
```diff
@@ -28,6 +95,41 @@ public DataFetcher<Double> lon() {
+
+  @Override
+  public DataFetcher<String> streetName() {
+    return environment -> getSource(environment).getStreetName().toString(environment.getLocale());
```
You could use `org.opentripplanner.ext.legacygraphqlapi.LegacyGraphQLUtils#getTranslation` if we want to enable fetching different translations here.
